File: src/main/java/ch/grassl/day03/impl/ItemImporter.java
package ch.grassl.day03.impl;

import ch.grassl.common.Importer;
import ch.grassl.util.ResourceReader;
import ch.grassl.util.Util;

import java.util.ArrayList;
import java.util.List;

public class ItemImporter implements Importer<Rucksack> {

    public static ItemImporter getInstance() {
        return new ItemImporter();
    }

    @Override
    public List<Rucksack> importData(String resource) {
        String[] lines = ResourceReader.of(resource).read();
        return mapToRucksacks(lines);
    }

    private List<Rucksack> mapToRucksacks(String[] lines) {
        List<Rucksack> rucksacks = new ArrayList<>();
        for (String line : lines) {
            if (!Util.isEmpty(line)) {
                rucksacks.add(mapToRucksack(line));
            }
        }
        return rucksacks;
    }

    private Rucksack mapToRucksack(String line) {
        String itemsFront = line.substring(0, line.length() / 2);
        String itemsBack = line.substring(line.length() / 2);
        return Rucksack.builder()
                .front(mapToItems(itemsFront))
                .back(mapToItems(itemsBack))
                .build();
    }

    private static List<Item> mapToItems(String items) {
        return items.chars()
                .mapToObj(c -> (char) c)
                .map(Item::new)
                .toList();
    }
}

File: src/main/java/ch/grassl/day03/impl/importer/ItemImporter.kt
package ch.grassl.day03.impl.importer

import ch.grassl.common.Importer
import ch.grassl.day03.impl.model.Item
import ch.grassl.day03.impl.model.Rucksack
import ch.grassl.util.ResourceReader

class ItemImporter : Importer<Rucksack> {

    override fun importData(resource: String?): List<Rucksack> {
        val rucksacks = ArrayList<Rucksack>()
        val lines = ResourceReader.of(resource).read()
        for (line in lines) {
            if (!line.isNullOrEmpty()) {
                rucksacks.add(
                    mapToRucksack(line)
                )
            }
        }
        return rucksacks
    }

    private fun mapToRucksack(line: String): Rucksack {
        val half = line.length / 2
        val itemsFront = line.substring(0, half)
        val itemsBack = line.substring(half)
        return Rucksack.Builder()
            .front(mapToItems(itemsFront))
            .back(mapToItems(itemsBack))
            .build()
    }

    private fun mapToItems(items: String): List<Item> {
        return items
            .map { Item(it) }
    }
}